frontend-react/src/components/EditableCompare.tsx

```diff
@@ -127,6 +127,19 @@ export const EditableCompare = forwardRef(
                       )
                     : textDifferMarkup(originalText, modifiedText);
 
+                // reformat the modified text in case it is not in correct JSON format
+                // then stick it back into the edited box. This ensures the highlight works correctly.
+                try {
+                    const reformatModifiedText = JSON.stringify(
+                        JSON.parse(modifiedText),
+                        null,
+                        2
+                    );
+                    setTextAreaContent(reformatModifiedText);
```

Comment:
Hmmm it's a bit odd to me that we're calling `setTextAreaContent` in `refreshDiffCallback`, which is already called immediately after `setTextAreaContent` in [`onChangeHandler`](https://github.com/CDCgov/prime-reportstream/blob/master/frontend-react/src/components/EditableCompare.tsx#L150-L151).  It miiiiiiight also be poor UX to replace what the user put in with a pretty-printed JSON (i.e., `JSON.stringify(whatever, null, 2)`) -- even though that'll guarantee the error demarcations are correct, it automatically changes the user input and prevents spaces and newlines from being added when the current JSON is valid:
![2022-12-13 11 21 19](https://user-images.githubusercontent.com/2421042/207387744-a8a35d06-376b-401c-acb7-1a4d8ae944d9.gif)

I think there's a bigger bug here due to the fact that the diffing is based on an indentation of two spaces (see [JsonSourceMap](https://github.com/CDCgov/prime-reportstream/blob/master/frontend-react/src/utils/DiffCompare/JsonSourceMap.ts#L86) and [JsonDiffer](https://github.com/CDCgov/prime-reportstream/blob/master/frontend-react/src/utils/DiffCompare/JsonDiffer.ts#L276-L277)), and user input may include less or more whitespace than that -- which is valid JSON as long as the structural pieces are all valid.

In my opinion, this is a fairly small UX bug that would require a much larger restructure to address so I'd be in favor of punting on it and coming back to it later if we need.  What are your thoughts?  (I'd also be in favor of looking into an external library to handle this so we don't have to maintain the logic ourselves, but that's a different conversation! :))

Reply:
> I'd also be in favor of looking into an external library to handle this so we don't have to maintain the logic ourselves, but that's a different conversation! :)

History: I investigated other solutions and was not able to find any that didn't come with their own set of issues. Also, the first iteration of the code was a simple text differ, you might see it referenced in places. It wasn't removed because it could still be handy for other data formats (FHIR?).

UI wise, is it possible to click directly on the Save button with invalid json in the edit area? 

One possible solutions that uses a fairly trivial UI change:
1. when the edit area gets focus, remove the highlighting (so you don't have to worry about keeping it in sync) and add a "check syntax" button somewhere (maybe overlay it in the upper left of the edit area).
2. When the "check syntax" is clicked, then apply formatting and syntax highlighting and syntax error checking and remove/disable the button.
3. Run "check syntax" function when the Save button is clicked to make sure there are no json syntax mistakes before calling the POST API.

Reply:
Thanks for the context, @TomNUSDS!  Yeah, I think moving it to a more user-initiated event like clicking a new 'Check Syntax' button would be better -- the issue was that the highlighting and reformatting was happening while the user could still be typing was a jarring experience so taking it out as its own action would solve this issue.  @penny-lischer what are your thoughts about this?

Reply:
Sure